scanpy/external/pl.py

```diff
@@ -234,3 +239,443 @@ def sam(
         if colorbar:
             plt.colorbar(cax, ax=axes)
     return axes
+
+
+def wishbone_marker_trajectory(
+    adata: AnnData, no_bins: int = 150, smoothing_factor: int = 1,
+):
+    """\
+    Plot marker trends along trajectory
+
+    Parameters
+    ----------
+    adata
+        Annotated data matrix.
+    no_bins
+        Number of bins for calculating marker density
+    smoothing_factor
+        Parameter controlling the degree of smoothing
+
+    Returns
+    -------
+    Updates `adata` with the following fields:
+
+    `weights_wishbone` : :class:`pandas.DataFrame` (`adata.uns`)
+        Computed gaussian weights for points at each location
+    `branch_point_bin` : 'int' (`adata.uns`)
+        Identifies the bin with the branch point. In case of no branching,
+        defaults to `no_bins`
+    `bins_wishbone` : :class:`numpy.ndarray` (`adata.uns`)
+        Computed bin locations and bin memberships
+    `branches_wishbone`: :class:`numpy.ndarray` (`adata.uns`)
+        In case of branching, returns a list of branches [2, 3].
+    """
+
+    # Compute bin locations and bin memberships
+    # Sort trajectory
+    trajectory = adata.obs['trajectory_wishbone'].sort_values()
+    bins = np.linspace(np.min(trajectory), np.max(trajectory), no_bins)
+
+    # Compute gaussian weights for points at each location
+    # Standard deviation estimated from Silverman's approximation
+    stdev = np.std(trajectory) * 1.34 * len(trajectory) ** (-1 / 5) * smoothing_factor
+    weights = np.exp(
+        -((np.tile(trajectory, [no_bins, 1]).T - bins) ** 2 / (2 * stdev ** 2))
+    ) * (1 / (2 * np.pi * stdev ** 2) ** 0.5)
+
+    # Adjust weights if data has branches
+    if 'branch_wishbone' in adata.obs.keys():
+        # Branch of the trunk
+        trunk = adata.obs['branch_wishbone'][trajectory.index[0]]
+        adata.uns['branches_wishbone'] = np.array([2, 3])
+        # Counts of branch cells in each bin
+        branch_counts = pd.DataFrame(np.zeros([len(bins) - 1, 3]), columns=[1, 2, 3])
+        for j in branch_counts.columns:
+            branch_counts[j] = pd.Series(
+                [
+                    pd.Series(
+                        adata.obs['branch_wishbone'][
+                            trajectory.index[
+                                (trajectory > bins[i - 1]) & (trajectory < bins[i])
+                            ]
+                        ]
+                        == j
+                    ).sum()
+                    for i in range(1, len(bins))
+                ]
+            )
```

Comment:
This looks weird! I think doing it the other way round should be computationally more efficient and prettier. Something like:

```py
branches = np.array([
    adata.obs['branch_wishbone'].values[trajectory.index[(trajectory > a) & (trajectory < b)]]
    for a, b in pairwise(bins)
])
# or axis=0, test what’s correct
branch_counts = np.array([(branches == i).sum(axis=1) for i in [1, 2, 3]])
```

Reply:
Good catch! again this is the original code adapted from Wishbone class. But, I will definitely test your suggestion and if it returns the expected result, I will change the code. Thank you for the tip!

Reply:
With the new change in the latest commit, this comment no longer relevant and can be resolved.